fuse_models/include/fuse_models/unicycle_2d_state_cost_function.h

```diff
@@ -0,0 +1,285 @@
+   *                         1 : yaw1 - First yaw
+   *                         2 : vel_linear1 - First linear velocity (array with x at index 0, y at index 1)
+   *                         3 : vel_yaw1 - First yaw velocity
+   *                         4 : acc_linear1 - First linear acceleration (array with x at index 0, y at index 1)
+   *                         5 : position2 - Second position (array with x at index 0, y at index 1)
+   *                         6 : yaw2 - Second yaw
+   *                         7 : vel_linear2 - Second linear velocity (array with x at index 0, y at index 1)
+   *                         8 : vel_yaw2 - Second yaw velocity
+   *                         9 : acc_linear2 - Second linear acceleration (array with x at index 0, y at index 1)
+   * @param[out] residual - The computed residual (error)
+   * @param[out] jacobians - Jacobians of the residuals wrt the parameters. Only computed if not NULL, and only
+   *                         computed for the parameters where jacobians[i] is not NULL.
+   * @return The return value indicates whether the computation of the residuals and/or jacobians was successful or not.
+   */
+  bool Evaluate(double const* const* parameters,
+                double* residuals,
+                double** jacobians) const override
+  {
+    double position_pred_x;
+    double position_pred_y;
+    double yaw_pred;
+    double vel_linear_pred_x;
+    double vel_linear_pred_y;
+    double vel_yaw_pred;
+    double acc_linear_pred_x;
+    double acc_linear_pred_y;
+    predict(
+      parameters[0][0],  // position1_x
+      parameters[0][1],  // position1_y
+      parameters[1][0],  // yaw1
+      parameters[2][0],  // vel_linear1_x
+      parameters[2][1],  // vel_linear1_y
+      parameters[3][0],  // vel_yaw1
+      parameters[4][0],  // acc_linear1_x
+      parameters[4][1],  // acc_linear1_y
+      dt_,
+      position_pred_x,
+      position_pred_y,
+      yaw_pred,
+      vel_linear_pred_x,
+      vel_linear_pred_y,
+      vel_yaw_pred,
+      acc_linear_pred_x,
+      acc_linear_pred_y,
+      jacobians);
+
+    residuals[0] = parameters[5][0] - position_pred_x;
+    residuals[1] = parameters[5][1] - position_pred_y;
+    residuals[2] = parameters[6][0] - yaw_pred;
+    residuals[3] = parameters[7][0] - vel_linear_pred_x;
+    residuals[4] = parameters[7][1] - vel_linear_pred_y;
+    residuals[5] = parameters[8][0] - vel_yaw_pred;
+    residuals[6] = parameters[9][0] - acc_linear_pred_x;
+    residuals[7] = parameters[9][1] - acc_linear_pred_y;
+
+    fuse_core::wrapAngle2D(residuals[2]);
+
+    // Scale the residuals by the square root information matrix to account for
+    // the measurement uncertainty.
+    Eigen::Map<fuse_core::Vector8d> residuals_map(residuals);
+    residuals_map.applyOnTheLeft(A_);
+
+    if (jacobians)
+    {
+      // It might be possible to simplify the code below implementing something like this but using compile-time
```

Comment:
This is a TODO comment. Something that could be explored, although it might require some helper `struct`s to implement the template recursion. Not sure if it'd be cleaner at the end. It'd be definitely cleaner here, but it'd require that template recursion magic somewhere, and it'd probably require access to `A_`, at least.

Reply:
I thought this was going to be easy...until I tried to write it up. Leaving this as a TODO item seems reasonable. It won't actually improve efficiency any; just reduce the amount of code duplication.